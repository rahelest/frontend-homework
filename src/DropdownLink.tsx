import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import { Company, ReduxState } from './types'
import { getIsDropdownMenuVisible, getSelectedCompany } from './selectors'
import { toggleDropdownMenuVisibility } from './actions'

import DropdownMenu from './DropdownMenu'
import useClickOutside from './useClickOutside'
import { useRef } from 'react'

type ReduxProps = {
  selectedCompany: Company
  isDropdownMenuVisible: boolean
}

type DispatchProps = {
  toggleDropdownMenuVisibility: () => void
}

export const DropdownLink = ({
  selectedCompany,
  isDropdownMenuVisible,
  toggleDropdownMenuVisibility,
}: ReduxProps & DispatchProps) => {
  const ref = useRef()
  useClickOutside(ref, isDropdownMenuVisible, toggleDropdownMenuVisibility)

  return (
    // @ts-ignore
    <div ref={ref}>
      <div className="nav__link" onClick={toggleDropdownMenuVisibility} data-test-nav-link>
        <div className="nav__link-text-wrapper">
          <div className="nav__link-text">Elon Musk</div>

          <div className="nav__link-subtext">{selectedCompany.name}</div>
        </div>

        <i className="material-icons-outlined nav__link-icon">settings</i>
      </div>

      {isDropdownMenuVisible && <DropdownMenu />}
    </div>
  )
}

export default connect(
  createStructuredSelector<ReduxState, ReduxProps>({
    // @ts-ignore
    selectedCompany: getSelectedCompany,
    isDropdownMenuVisible: getIsDropdownMenuVisible,
  }),
  { toggleDropdownMenuVisibility },
)(DropdownLink)
