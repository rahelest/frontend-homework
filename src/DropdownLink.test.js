import { shallow } from 'enzyme'
import { DropdownLink } from './DropdownLink'

describe('<DropdownLink />', () => {
  const toggleDropdownMenuVisibilityMock = jest.fn()

  const render = (props) =>
    shallow(
      <DropdownLink
        selectedCompany={{ id: 1, name: 'Dummy company' }}
        toggleDropdownMenuVisibility={toggleDropdownMenuVisibilityMock}
        {...props}
      />,
    )

  it('renders without dropdown menu', () => {
    expect(render({ isDropdownMenuVisible: false })).toMatchSnapshot()
  })

  it('renders with dropdown menu', () => {
    expect(render({ isDropdownMenuVisible: true })).toMatchSnapshot()
  })

  it('calls toggleDropdownMenuVisibility when nav link is clicked', () => {
    const component = render({ isDropdownMenuVisible: false })
    component.find('[data-test-nav-link]').simulate('click')

    expect(toggleDropdownMenuVisibilityMock).toHaveBeenCalled()
  })
})
