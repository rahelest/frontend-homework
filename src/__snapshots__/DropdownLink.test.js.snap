// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownLink /> renders with dropdown menu 1`] = `
<div>
  <div
    className="nav__link"
    data-test-nav-link={true}
    onClick={[MockFunction]}
  >
    <div
      className="nav__link-text-wrapper"
    >
      <div
        className="nav__link-text"
      >
        Elon Musk
      </div>
      <div
        className="nav__link-subtext"
      >
        Dummy company
      </div>
    </div>
    <i
      className="material-icons-outlined nav__link-icon"
    >
      settings
    </i>
  </div>
  <DropdownMenu />
</div>
`;

exports[`<DropdownLink /> renders without dropdown menu 1`] = `
<div>
  <div
    className="nav__link"
    data-test-nav-link={true}
    onClick={[MockFunction]}
  >
    <div
      className="nav__link-text-wrapper"
    >
      <div
        className="nav__link-text"
      >
        Elon Musk
      </div>
      <div
        className="nav__link-subtext"
      >
        Dummy company
      </div>
    </div>
    <i
      className="material-icons-outlined nav__link-icon"
    >
      settings
    </i>
  </div>
</div>
`;
